Tests/SessionTests.swift

```diff
@@ -1123,20 +1153,262 @@ class SessionTestCase: BaseTestCase {
+        XCTAssertEqual(handler.adaptedCount, 1)
+        XCTAssertEqual(handler.retryCalledCount, 4)
+        XCTAssertEqual(handler.retryCount, 4)
+        XCTAssertEqual(request.retryCount, 1)
+        XCTAssertEqual(json1Response?.result.isSuccess, false)
+        XCTAssertEqual(json2Response?.result.isSuccess, false)
+        XCTAssertTrue(session.requestTaskMap.isEmpty)
+
+        let errors: [AFError] = [json1Response, json2Response].compactMap { $0?.error?.asAFError }
+        XCTAssertEqual(errors.count, 2)
+
+        for error in errors {
+            XCTAssertTrue(error.isRequestAdaptationError)
+            XCTAssertEqual(error.localizedDescription, "Request adaption failed with error: URL is not valid: /adapt/error/2")
+        }
+    }
+
+    func testThatSessionCallsResponseSerializerCompletionsWhenAdapterThrowsErrorDuringRetryForDownloads() {
+        // Four retries should occur given this scenario:
+        // 1) Retrier is called from first response serializer failure (trips retry)
+        // 2) Retrier is called by Session for adapt error thrown
+        // 3) Retrier is called again from first response serializer failure
+        // 4) Retrier is called from second response serializer failure
+
+        // Given
+        let handler = RequestHandler()
+        handler.throwsErrorOnSecondAdapt = true
+
+        let session = Session()
+
+        let json1Expectation = self.expectation(description: "request should eventually fail")
+        var json1Response: DownloadResponse<Any>?
+
+        let json2Expectation = self.expectation(description: "request should eventually fail")
+        var json2Response: DownloadResponse<Any>?
+
+        // When
+        let request = session.download("https://httpbin.org/image/jpeg", interceptor: handler)
+            .validate()
+            .responseJSON { response in
+                json1Response = response
+                json1Expectation.fulfill()
+            }
+            .responseJSON { response in
+                json2Response = response
+                json2Expectation.fulfill()
+            }
+
+        waitForExpectations(timeout: 10, handler: nil)
+
+        // Then
+        XCTAssertEqual(handler.adaptCalledCount, 2)
+        XCTAssertEqual(handler.adaptedCount, 1)
+        XCTAssertEqual(handler.retryCalledCount, 4)
+        XCTAssertEqual(handler.retryCount, 4)
+        XCTAssertEqual(request.retryCount, 1)
+        XCTAssertEqual(json1Response?.result.isSuccess, false)
+        XCTAssertEqual(json2Response?.result.isSuccess, false)
+        XCTAssertTrue(session.requestTaskMap.isEmpty)
+
+        let errors: [AFError] = [json1Response, json2Response].compactMap { $0?.error?.asAFError }
+        XCTAssertEqual(errors.count, 2)
+
+        for error in errors {
+            XCTAssertTrue(error.isRequestAdaptationError)
+            XCTAssertEqual(error.localizedDescription, "Request adaption failed with error: URL is not valid: /adapt/error/2")
+        }
+    }
```

Comment:
These tests should cover all the use cases I can think of. Everything is working exactly as expected. 👍🏻